### ISSUE TITLE
Load pretrain model

### ISSUE THREAD

**Issue by leezonpen** [NVIDIA/apex#213]
Hi, I use DDP() and the load checkpoint. It seem like that the parameters didn't work. Should i put loading checkout before DDP()? Or need i broadcast model parameters loaded from pretrain-model?

**Comment by mcarilli** [comment]
The [imagenet example](https://github.com/NVIDIA/apex/tree/master/examples/imagenet) shows checkpointing with DDP, maybe that will help.  I'm working on giving Amp more thorough support for opt_level-aware checkpointing, and I will create some more focused examples in the process.  I probably won't get a chance to work on this until the end of next week, though, because I have other features to implement as well.

**Comment by glample** [comment]
@mcarilli is the imagenet example working? I'm doing something very similar, but when I reload a pretrained model (trained with AMP and 02 optimization), the loss of my model is absolutely huge, something is wrong. When I evaluate the reloaded model though, I observe the expected results, so it seems that the reloading issue is only in the optimizer. I tried to reload just the model from my checkpoints, and ignore the optimizer, but it's not optimal since I use the Adam optimizer, so when I do this my loss increases a little bit compared to just before the checkpoint.

I have no issue when I use FP16_optimizer though. Any ideas of what can be the issue, or a temporary fix? Given that the optimizer state_dict is entirely reloaded, I really don't see what could go wrong, only thing I can think of would be that some information about the optimizer is not stored in the state dict.

**Comment by mcarilli** [comment]
Are you using FusedAdam, or torch.optim.Adam?  If you are using FusedAdam, can you try torch.optim.Adam instead, and see if the problem persists?  That would help us sandbox out some potential issues.

**Comment by glample** [comment]
No unfortunately I'm already using torch.optim.Adam. In case it helps, I also noticed that with O1 optimization the problem does not happen (sorry for not being able to provide a simple script to reproduce the issue).

**Comment by mcarilli** [comment]
I think it's because O2 uses explicit master weights, which are not being saved and restored properly.  We're aware of this and we're working on a solution for checkpointing that should enable all opt_levels to be saved and restored.  I'll ping this issue when it's merged.

**Comment by felix-schneider** [comment]
Is there an update on this? Does this issue still exist?

**Comment by glample** [comment]
Yes it is, c.f. https://github.com/NVIDIA/apex/issues/250

**Comment by ptrblck** [comment]
@leezonpen, @felix-schneider 
Checkpointing just got merged into out master branch.
Checkout the [README](https://github.com/NVIDIA/apex#checkpointing) to see an example usage.

**Comment by voldemortX** [comment]
> @felix-schneider
> Checkpointing just got merged into out master branch.
> Checkout the [README](https://github.com/NVIDIA/apex#checkpointing) to see an example usage.

What if I'm using a pre-trained model that doesn't have a amp.state_dict()? e.g. ImageNet pre-trained without amp. Should I still load it latter than amp.initialize() or before? (using level O1)

**Comment by ptrblck** [comment]
@voldemortX I would stick to the ImageNet example, i.e. load the `state_dict` after calling `amp.initialize`.
